// traverse-async/lib/traverse-async.js
// Generated by CoffeeScript 1.9.2
(function() {
  var async, config, isArr, isObj;

  async = require('async');

  isObj = function(obj) {
    return '[object Object]' === Object.prototype.toString.call(obj);
  };

  isArr = function(arr) {
    return '[object Array]' === Object.prototype.toString.call(arr);
  };

  config = {
    parallel: 1
  };

  exports.config = function(obj) {
    var key, results, val;
    results = [];
    for (key in obj) {
      val = obj[key];
      results.push(config[key] = val);
    }
    return results;
  };

  exports.traverse = function(data, userCallback, done) {
    var q, traverseNode;
    traverseNode = function(context, next) {
      var i, key, len, ref;
      if ((isArr(context.node) || isObj(context.node)) && typeof context.node.then !== "function") {
        ref = Object.keys(context.node);
        for (i = 0, len = ref.length; i < len; i++) {
          key = ref[i];
          q.push({
            parent: context.node,
            node: context.node[key],
            key: key,
            path: context.path.concat([key])
          });
        }
      }
      return userCallback.call(context, context.node, next);
    };
    q = async.queue(traverseNode, config.parallel);
    q.drain = function() {
      if (done) {
        return done(data);
      }
    };
    q.push({
      node: data,
      path: [],
      isRoot: true
    });
    return {
      "break": q.kill,
      push: q.push
    };
  };

}).call(this);
